src/array/cpu/array_utils.h

```diff
@@ -30,10 +30,15 @@ class IdHashMap {
 
   // Construct the hashmap using the given id array.
   // The id array could contain duplicates.
+  // If the id array has no duplicates, the array will be relabeled to consecutive
+  // integers starting from 0.
```

Comment:
For hashmap. If it is expected to be a little bit large. It is better to preallocate the space for the hashtable other than let it grow during insertions.